src/device_manager.rs

```diff
@@ -0,0 +1,365 @@
+    /// VMM is responsible for providing the allocated resources to virtual device.
+    ///
+    /// # Arguements
+    ///
+    /// * `device`: device instance object to be registered
+    /// * `resources`: resources that this device owns, might include
+    ///                port I/O and memory-mapped I/O ranges, irq number, etc.
+    pub fn register_device_io(
+        &mut self,
+        device: Arc<dyn DeviceIo>,
+        resources: &[Resource],
+    ) -> Result<()> {
+        // Register and mark device resources
+        // The resources addresses being registered are sucessfully allocated before.
+        for (idx, res) in resources.iter().enumerate() {
+            match *res {
+                Resource::PioAddressRange { base, size } => {
+                    if self
+                        .pio_bus
+                        .insert(IoRange::new_pio_range(base, size), device.clone())
+                        .is_some()
+                    {
+                        // Unregister registered resources.
+                        self.unregister_device_io(&resources[0..idx])?;
+
+                        return Err(Error::DeviceOverlap);
+                    }
+                }
+                Resource::MmioAddressRange { base, size } => {
+                    if self
+                        .mmio_bus
+                        .insert(IoRange::new_mmio_range(base, size), device.clone())
+                        .is_some()
+                    {
+                        // Unregister registered resources.
+                        self.unregister_device_io(&resources[0..idx])?;
+
+                        return Err(Error::DeviceOverlap);
+                    }
+                }
+                _ => continue,
+            }
+        }
+        Ok(())
+    }
+
+    /// Unregister a device from `IoManager`, e.g. users specified removing.
+    /// VMM pre-fetches the resources e.g. dev.get_assigned_resources()
+    /// VMM is responsible for freeing the resources.
+    ///
+    /// # Arguements
+    ///
+    /// * `resources`: resources that this device owns, might include
+    ///                port I/O and memory-mapped I/O ranges, irq number, etc.
+    pub fn unregister_device_io(&mut self, resources: &[Resource]) -> Result<()> {
+        for res in resources.iter() {
+            match *res {
+                Resource::PioAddressRange { base, size } => {
+                    self.pio_bus.remove(&IoRange::new_pio_range(base, size));
+                }
+                Resource::MmioAddressRange { base, size } => {
+                    self.mmio_bus.remove(&IoRange::new_mmio_range(base, size));
+                }
+                _ => continue,
+            }
+        }
+        Ok(())
+    }
+
+    fn first_before(
+        &self,
+        addr: IoAddress,
+        map: BTreeMap<IoRange, Arc<dyn DeviceIo>>,
+    ) -> Option<(IoRange, Arc<dyn DeviceIo>)> {
+        for (range, dev) in map.iter().rev() {
```

Comment:
We should use BtreeMap::range()(https://doc.rust-lang.org/std/collections/struct.BTreeMap.html#method.range) here, it's available since rust 1.17 and should be more efficient than iter().

Reply:
> We should use BtreeMap::range()(https://doc.rust-lang.org/std/collections/struct.BTreeMap.html#method.range) here, it's available since rust 1.17 and should be more efficient than iter().

Updated with new method.

Reply:
It seems our CI doesn't accept `std::iter::DoubleEndedIterator` trait which is implemented for `Range` used by `BtreeMap::range()`? @sameo @jiangliu